jupyterhub/handlers/base.py

```diff
@@ -1159,16 +1159,36 @@ def spawn_home_error(self):
             "<a href='{home}'>home page</a>.".format(home=home)
         )
 
-    def get_template(self, name):
-        """Return the jinja template object for a given name"""
-        return self.settings['jinja2_env'].get_template(name)
+    def get_template(self, name, sync=False):
+        """
+        Return the jinja template object for a given name
+
+        If sync is True, we return a Template that is compiled without async support.
+        Only those can be used in synchronous code.
+
+        If sync is False, we return a Template that is compiled with async support
+        """
+        if sync:
+            key = 'jinja2_env_sync'
+        else:
+            key = 'jinja2_env'
+        return self.settings[key].get_template(name)
+
+    def render_template(self, name, sync=False, **ns):
```

Comment:
If we want to preserve backward-compatibility, we can default to `sync=True`. However, this backward-compatibility concern only applies to third-party template-based pages. These are *technically* possible, but I wouldn't call them "supported" since I specifically *do not* want JupyterHub customizations to add extra pages like notebook extensions do. Such extensions really do belong outside the Hub as services.